### PR TITLE
feat: show driver summary

### PR DESCRIPTION
**Pull request by Graborg**

### CODE REVIEW

**Comment by Graborg** [comment]
Just need to fix the types in the test

**Comment by Graborg** [comment]
[Send a summary over the plan to the driver when they login / after plan is confirmed](https://trello.com/c/Ovzkefk7/309-send-a-summary-over-the-plan-to-the-driver-when-they-login-after-plan-is-confirmed)

File: packages/driver-interface/package.json [hunk 1]
```diff
@@ -30,11 +30,14 @@
     "@types/dotenv": "^8.2.0",
     "@types/express": "^4.17.8",
     "@types/ioredis": "^4.17.5",
+    "@types/jest": "^26.0.15",
```
Comment (lundeeell) [inline packages/driver-interface/package.json:33]: Missing updated package-lock.json

File: packages/driver-interface/src/services/bot.ts [hunk 2]
```diff
@@ -5,10 +5,24 @@ import { v4 as uuid } from 'uuid'
...
+export const driverIsLoggedIn = async (
+  telegramId: number
+): Promise<boolean> => {
+  const vehicleId = await cache.getVehicleIdByTelegramId(telegramId)
+  return !!vehicleId
+}
+
+export const onInstructionsReceived = async (
+  telegramId: number,
+  instructions: Instruction[][]
```
Comment (lundeeell) [inline packages/driver-interface/src/services/bot.ts:19]: Suggestion: In most places, Instruction[][] is named instructionGroups. Goes for many places :)

File: packages/driver-interface/src/services/bot.ts [hunk 3]
```diff
@@ -5,10 +5,24 @@ import { v4 as uuid } from 'uuid'
...
+): Promise<Message> => {
```
Comment (lundeeell) [inline packages/driver-interface/src/services/bot.ts:20]:
```suggestion
): Promise<Message | void> => {
```

File: packages/driver-interface/src/services/google.ts [hunk 4]
```diff
@@ -35,13 +35,21 @@ export const getAddressFromCoordinate = (coordinate: {
...
-    'https://www.google.com/maps/dir'
-  )
+export const getDirectionsFromInstructions = (
+  instructionGroups: Instruction[][]
+) =>
+  instructionGroups
+    .flatMap((instructionGroup: Instruction[]) =>
+      instructionGroup.map(({ address }) => address)
+    )
+    .filter(
```
Comment (lundeeell) [inline packages/driver-interface/src/services/google.ts:45]: Basically the same thing is done in convertInstructionsToSummaryList, maybe they could share base functionality, as Instruction[] are grouped by address.

File: packages/driver-interface/src/services/messaging.ts [hunk 5]
```diff
@@ -34,23 +36,51 @@ export const onDriverLoginSuccessful = (
     )
   )
 
+export const sendSummary = (
+  telegramId: number,
+  instructions: Instruction[][]
```
Comment (lundeeell) [inline packages/driver-interface/src/services/messaging.ts:41]: Same as before, often named instructionGroups, to not confuse with Instruction[]

**Review by lundeeell — CHANGES_REQUESTED** [review]
Mostly suggestions, but at least missing updated package-lock.json

**Comment by lundeeell** [comment]
And also, I didn't get the summery on login, but can be how I tested. Please double check if that is supposed to happen!

File: packages/driver-interface/src/services/bot.ts [hunk 6]
```diff
@@ -56,36 +70,34 @@ export const handleNextDriverInstruction = async (
...
-
-    const [currentInstructionGroup] = instructions
-
-    if (!currentInstructionGroup) {
+    } else if (!instructions.length) {
+      console.log('Vehicle finished all instructions:', vehicleId)
       cache.setInstructions(vehicleId, null)
       return messaging.sendDriverFinishedMessage(telegramId)
-    }
+    } else {
```
[the remark this answers is not in the document]
Comment (Graborg) [reply packages/driver-interface/src/services/bot.ts:79]: 👍

File: packages/driver-interface/src/services/google.ts (hunk 4 above)
Comment (Graborg) [reply packages/driver-interface/src/services/google.ts:45]: Don't see what you mean here.

**Comment by Graborg** [comment]
To test the flow:
1. Create a vehicle with a telefonnumber which is the same as in your telegram.
2. Create a booking for the vehicle to pick up

3a. Confirm the plan
4a. Login in telegram

3b. Login in telegram
4b. Confirm the plan

You should now get instructions in telegram

**Review by lundeeell — APPROVED** [review]
👍

**Pull request merged**